src/Adapter/Product/Update/ProductShopUpdater.php

```diff
@@ -235,7 +236,15 @@ private function copyCarriersToShop(Product $sourceProduct, ShopId $targetShopId
     private function copyImageAssociations(ProductId $productId, ShopId $sourceShopId, ShopId $targetShopId): void
     {
         $imagesFromSourceShop = $this->productImageMultiShopRepository->getImages($productId, ShopConstraint::shop($sourceShopId->getValue()));
+        $targetImageIds = array_map(static function (ImageId $imageId): int {
+            return $imageId->getValue();
+        }, $this->productImageMultiShopRepository->getImagesIds($productId, ShopConstraint::shop($targetShopId->getValue())));
+
         foreach ($imagesFromSourceShop as $image) {
+            // skip image if it is already associated with the target shop
+            if (in_array((int) $image->id, $targetImageIds, true)) {
```

Comment:
these changes are also not mandatory, as the root issue was that the images wasn't being deleted from shops when product was deleted, but I think its still nice to have in case for some reason some images are left, so we don't get the sql UniqueConstraint error.